File: src/secrets/secret-manager.ts
/**
 * Secret Manager Interface
 *
 * Provides a unified interface for managing secrets from multiple backends.
 * Supports pluggable backends (Vault, AWS Secrets Manager, etc.) with fallback chains.
 */

export interface SecretBackend {
  /**
   * Name of the backend
   */
  name: string;

  /**
   * Whether this backend is currently available/authenticated
   */
  isAvailable(): Promise<boolean>;

  /**
   * Retrieve a secret value
   */
  getSecret(key: string): Promise<string | null>;

  /**
   * Store a secret value
   */
  putSecret(key: string, value: string): Promise<void>;

  /**
   * Delete a secret
   */
  deleteSecret(key: string): Promise<void>;

  /**
   * List all available secret keys
   */
  listSecrets(pattern?: string): Promise<string[]>;

  /**
   * Rotate a secret (if supported)
   */
  rotateSecret?(key: string): Promise<string>;

  /**
   * Get audit log entries (if supported)
   */
  getAuditLog?(filter?: AuditFilter): Promise<AuditEntry[]>;

  /**
   * Close connection and cleanup
   */
  close(): Promise<void>;
}

export interface AuditFilter {
  startTime?: Date;
  endTime?: Date;
  action?: 'get' | 'put' | 'delete' | 'rotate';
  key?: string;
  user?: string;
}

export interface AuditEntry {
  timestamp: Date;
  action: 'get' | 'put' | 'delete' | 'rotate';
  key: string;
  status: 'success' | 'failed';
  user?: string;
  error?: string;
}

export interface SecretManagerConfig {
  backends: SecretBackend[];
  failover?: boolean; // Try next backend on failure
  cache?: CacheConfig;
  audit?: boolean;
}

export interface CacheConfig {
  enabled: boolean;
  ttl: number; // milliseconds
  maxSize: number; // max secrets to cache
}

/**
 * SecretManager - Manages secrets across multiple backends with fallback support
 */
export class SecretManager {
  private backends: SecretBackend[];
  private failover: boolean;
  private cache: Map<string, CachedSecret> = new Map();
  private cacheConfig?: CacheConfig;
  private auditLog: AuditEntry[] = [];
  private auditEnabled: boolean;

  constructor(config: SecretManagerConfig) {
    this.backends = config.backends;
    this.failover = config.failover ?? true;
    this.cacheConfig = config.cache;
    this.auditEnabled = config.audit ?? true;
  }

  /**
   * Get a secret from the first available backend
   */
  async getSecret(key: string): Promise<string | null> {
    // Check cache first
    if (this.cacheConfig?.enabled) {
      const cached = this.cache.get(key);
      if (cached && !this.isCacheExpired(cached)) {
        this.recordAudit('get', key, 'success');
        return cached.value;
      }
    }

    // Try each backend
    let backendsAvailable = false;
    for (const backend of this.backends) {
      try {
        if (!(await backend.isAvailable())) continue;
        backendsAvailable = true;

        const value = await backend.getSecret(key);
        if (value !== null) {
          // Cache the result
          if (this.cacheConfig?.enabled) {
            this.setCacheEntry(key, value);
          }
          this.recordAudit('get', key, 'success');
          return value;
        }
      } catch (error) {
        this.recordAudit('get', key, 'failed', (error as Error).message);
        if (!this.failover) throw error;
        continue;
      }
    }

    // If at least one backend was available, return null (key not found)
    if (backendsAvailable) {
      this.recordAudit('get', key, 'success');
      return null;
    }

    // No backends were available
    throw new Error(`Secret '${key}' not found in any backend`);
  }

  /**
   * Store a secret in all configured backends
   */
  async putSecret(key: string, value: string): Promise<void> {
    const errors: Error[] = [];

    for (const backend of this.backends) {
      try {
        if (!(await backend.isAvailable())) continue;
        await backend.putSecret(key, value);
        this.recordAudit('put', key, 'success');
      } catch (error) {
        this.recordAudit('put', key, 'failed', (error as Error).message);
        errors.push(error as Error);
        if (!this.failover) throw error;
      }
    }

    // Update cache
    if (this.cacheConfig?.enabled) {
      this.setCacheEntry(key, value);
    }

    if (errors.length === this.backends.length) {
      throw new Error(`Failed to store secret '${key}' in all backends`);
    }
  }

  /**
   * Delete a secret from all backends
   */
  async deleteSecret(key: string): Promise<void> {
    this.cache.delete(key);
    const errors: Error[] = [];

    for (const backend of this.backends) {
      try {
        if (!(await backend.isAvailable())) continue;
        await backend.deleteSecret(key);
        this.recordAudit('delete', key, 'success');
      } catch (error) {
        this.recordAudit('delete', key, 'failed', (error as Error).message);
        errors.push(error as Error);
        if (!this.failover) throw error;
      }
    }

    if (errors.length === this.backends.length) {
      throw new Error(`Failed to delete secret '${key}' from all backends`);
    }
  }

  /**
   * List secrets from the first available backend
   */
  async listSecrets(pattern?: string): Promise<string[]> {
    for (const backend of this.backends) {
      try {
        if (!(await backend.isAvailable())) continue;
        return await backend.listSecrets(pattern);
      } catch (error) {
        if (!this.failover) throw error;
        continue;
      }
    }

    throw new Error('No backends available for listing secrets');
  }

  /**
   * Rotate a secret (if supported by backend)
   */
  async rotateSecret(key: string): Promise<string> {
    for (const backend of this.backends) {
      try {
        if (!(await backend.isAvailable())) continue;
        if (!backend.rotateSecret) continue;

        const newValue = await backend.rotateSecret(key);
        this.cache.delete(key);
        this.recordAudit('rotate', key, 'success');
        return newValue;
      } catch (error) {
        this.recordAudit('rotate', key, 'failed', (error as Error).message);
        if (!this.failover) throw error;
        continue;
      }
    }

    throw new Error(`Secret '${key}' rotation not supported in any backend`);
  }

  /**
   * Get audit log entries
   */
  getAuditLog(filter?: AuditFilter): AuditEntry[] {
    return this.auditLog.filter((entry) => {
      if (filter?.startTime && entry.timestamp < filter.startTime) return false;
      if (filter?.endTime && entry.timestamp > filter.endTime) return false;
      if (filter?.action && entry.action !== filter.action) return false;
      if (filter?.key && entry.key !== filter.key) return false;
      if (filter?.user && entry.user !== filter.user) return false;
      return true;
    });
  }

  /**
   * Clear audit log
   */
  clearAuditLog(): void {
    this.auditLog = [];
  }

  /**
   * Close all backends
   */
  async close(): Promise<void> {
    await Promise.all(this.backends.map((b) => b.close()));
    this.cache.clear();
  }

  /**
   * Get cache statistics
   */
  getCacheStats() {
    return {
      size: this.cache.size,
      maxSize: this.cacheConfig?.maxSize ?? 0,
      ttl: this.cacheConfig?.ttl ?? 0,
      entries: Array.from(this.cache.entries()).map(([key, cached]) => ({
        key,
        age: Date.now() - cached.timestamp,
        expired: this.isCacheExpired(cached),
      })),
    };
  }

  /**
   * Clear cache
   */
  clearCache(): void {
    this.cache.clear();
  }

  // Private helpers

  private setCacheEntry(key: string, value: string): void {
    if (!this.cacheConfig?.enabled) return;

    // Evict if cache is full
    if (
      this.cacheConfig.maxSize &&
      this.cache.size >= this.cacheConfig.maxSize
    ) {
      const firstKey = this.cache.keys().next().value as string;
      if (firstKey) {
        this.cache.delete(firstKey);
      }
    }

    this.cache.set(key, {
      value,
      timestamp: Date.now(),
    });
  }

  private isCacheExpired(cached: CachedSecret): boolean {
    if (!this.cacheConfig?.ttl) return false;
    return Date.now() - cached.timestamp > this.cacheConfig.ttl;
  }

  private recordAudit(
    action: 'get' | 'put' | 'delete' | 'rotate',
    key: string,
    status: 'success' | 'failed',
    error?: string
  ): void {
    if (!this.auditEnabled) return;

    this.auditLog.push({
      timestamp: new Date(),
      action,
      key,
      status,
      error,
      user: process.env.USER,
    });

    // Keep audit log size bounded (last 1000 entries)
    if (this.auditLog.length > 1000) {
      this.auditLog = this.auditLog.slice(-1000);
    }
  }
}

interface CachedSecret {
  value: string;
  timestamp: number;
}

export default SecretManager;

File: python/ter/secrets/secret_manager.py
"""
Secret Manager for TER (Python Implementation)

Provides a unified interface for managing secrets from multiple backends.
"""

from abc import ABC, abstractmethod
from typing import Dict, List, Optional, Any
from datetime import datetime
from dataclasses import dataclass, field


@dataclass
class AuditFilter:
    """Filter for audit log queries"""
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    action: Optional[str] = None  # 'get', 'put', 'delete', 'rotate'
    key: Optional[str] = None
    user: Optional[str] = None


@dataclass
class AuditEntry:
    """Audit log entry"""
    timestamp: datetime
    action: str  # 'get', 'put', 'delete', 'rotate'
    key: str
    status: str  # 'success', 'failed'
    user: Optional[str] = None
    error: Optional[str] = None


@dataclass
class CacheConfig:
    """Cache configuration"""
    enabled: bool = True
    ttl: int = 5000  # milliseconds
    max_size: int = 100  # max secrets to cache


class SecretBackend(ABC):
    """Abstract base class for secret backends"""

    @property
    @abstractmethod
    def name(self) -> str:
        """Name of the backend"""
        pass

    @abstractmethod
    async def is_available(self) -> bool:
        """Check if backend is available"""
        pass

    @abstractmethod
    async def get_secret(self, key: str) -> Optional[str]:
        """Retrieve a secret"""
        pass

    @abstractmethod
    async def put_secret(self, key: str, value: str) -> None:
        """Store a secret"""
        pass

    @abstractmethod
    async def delete_secret(self, key: str) -> None:
        """Delete a secret"""
        pass

    @abstractmethod
    async def list_secrets(self, pattern: Optional[str] = None) -> List[str]:
        """List secrets"""
        pass

    async def rotate_secret(self, key: str) -> str:
        """Rotate a secret (optional, may not be supported by all backends)"""
        raise NotImplementedError(f'{self.name} does not support secret rotation')

    async def get_audit_log(self, filter: Optional[AuditFilter] = None) -> List[AuditEntry]:
        """Get audit log entries (optional, may not be supported by all backends)"""
        return []

    @abstractmethod
    async def close(self) -> None:
        """Close connection and cleanup"""
        pass


@dataclass
class _CachedSecret:
    """Internal cached secret wrapper"""
    value: str
    timestamp: int


class SecretManager:
    """Manages secrets across multiple backends with fallback support"""

    def __init__(
        self,
        backends: List[SecretBackend],
        failover: bool = True,
        cache: Optional[CacheConfig] = None,
        audit: bool = True,
    ):
        """Initialize SecretManager"""
        self.backends = backends
        self.failover = failover
        self.cache_config = cache or CacheConfig()
        self.audit_enabled = audit
        self.cache: Dict[str, _CachedSecret] = {}
        self.audit_log: List[AuditEntry] = []

    async def get_secret(self, key: str) -> str:
        """Get a secret from the first available backend"""
        import time

        # Check cache first
        if self.cache_config.enabled:
            cached = self.cache.get(key)
            if cached and not self._is_cache_expired(cached):
                self._record_audit('get', key, 'success')
                return cached.value

        # Try each backend
        for backend in self.backends:
            try:
                if not await backend.is_available():
                    continue

                value = await backend.get_secret(key)
                if value is not None:
                    # Cache the result
                    if self.cache_config.enabled:
                        self._set_cache_entry(key, value)
                    self._record_audit('get', key, 'success')
                    return value
            except Exception as e:
                self._record_audit('get', key, 'failed', str(e))
                if not self.failover:
                    raise
                continue

        raise RuntimeError(f"Secret '{key}' not found in any backend")

    async def put_secret(self, key: str, value: str) -> None:
        """Store a secret in all configured backends"""
        errors = []

        for backend in self.backends:
            try:
                if not await backend.is_available():
                    continue
                await backend.put_secret(key, value)
                self._record_audit('put', key, 'success')
            except Exception as e:
                self._record_audit('put', key, 'failed', str(e))
                errors.append(e)
                if not self.failover:
                    raise

        # Update cache
        if self.cache_config.enabled:
            self._set_cache_entry(key, value)

        if errors and len(errors) == len(self.backends):
            raise RuntimeError(f"Failed to store secret '{key}' in all backends")

    async def delete_secret(self, key: str) -> None:
        """Delete a secret from all backends"""
        self.cache.pop(key, None)
        errors = []

        for backend in self.backends:
            try:
                if not await backend.is_available():
                    continue
                await backend.delete_secret(key)
                self._record_audit('delete', key, 'success')
            except Exception as e:
                self._record_audit('delete', key, 'failed', str(e))
                errors.append(e)
                if not self.failover:
                    raise

        if errors and len(errors) == len(self.backends):
            raise RuntimeError(f"Failed to delete secret '{key}' from all backends")

    async def list_secrets(self, pattern: Optional[str] = None) -> List[str]:
        """List secrets from the first available backend"""
        for backend in self.backends:
            try:
                if not await backend.is_available():
                    continue
                return await backend.list_secrets(pattern)
            except Exception:
                if not self.failover:
                    raise
                continue

        raise RuntimeError('No backends available for listing secrets')

    async def rotate_secret(self, key: str) -> str:
        """Rotate a secret"""
        for backend in self.backends:
            try:
                if not await backend.is_available():
                    continue

                new_value = await backend.rotate_secret(key)
                self.cache.pop(key, None)
                self._record_audit('rotate', key, 'success')
                return new_value
            except NotImplementedError:
                continue
            except Exception as e:
                self._record_audit('rotate', key, 'failed', str(e))
                if not self.failover:
                    raise
                continue

        raise RuntimeError(f"Secret '{key}' rotation not supported in any backend")

    def get_audit_log(self, filter: Optional[AuditFilter] = None) -> List[AuditEntry]:
        """Get audit log entries"""
        entries = self.audit_log

        if filter:
            if filter.start_time:
                entries = [e for e in entries if e.timestamp >= filter.start_time]
            if filter.end_time:
                entries = [e for e in entries if e.timestamp <= filter.end_time]
            if filter.action:
                entries = [e for e in entries if e.action == filter.action]
            if filter.key:
                entries = [e for e in entries if e.key == filter.key]
            if filter.user:
                entries = [e for e in entries if e.user == filter.user]

        return entries

    def clear_audit_log(self) -> None:
        """Clear audit log"""
        self.audit_log = []

    async def close(self) -> None:
        """Close all backends"""
        for backend in self.backends:
            await backend.close()
        self.cache.clear()

    def get_cache_stats(self) -> Dict[str, Any]:
        """Get cache statistics"""
        return {
            'size': len(self.cache),
            'max_size': self.cache_config.max_size,
            'ttl': self.cache_config.ttl,
            'entries': [
                {
                    'key': key,
                    'age': self._current_time_ms() - cached.timestamp,
                    'expired': self._is_cache_expired(cached),
                }
                for key, cached in self.cache.items()
            ],
        }

    def clear_cache(self) -> None:
        """Clear cache"""
        self.cache.clear()

    # Private helpers

    def _set_cache_entry(self, key: str, value: str) -> None:
        """Set a cache entry with LRU eviction"""
        if not self.cache_config.enabled:
            return

        # Evict if cache is full
        if self.cache_config.max_size and len(self.cache) >= self.cache_config.max_size:
            first_key = next(iter(self.cache))
            del self.cache[first_key]

        self.cache[key] = _CachedSecret(
            value=value,
            timestamp=self._current_time_ms(),
        )

    def _is_cache_expired(self, cached: _CachedSecret) -> bool:
        """Check if cache entry has expired"""
        if not self.cache_config.ttl:
            return False
        return self._current_time_ms() - cached.timestamp > self.cache_config.ttl

    def _record_audit(
        self,
        action: str,
        key: str,
        status: str,
        error: Optional[str] = None,
    ) -> None:
        """Record audit log entry"""
        if not self.audit_enabled:
            return

        import os

        entry = AuditEntry(
            timestamp=datetime.now(),
            action=action,
            key=key,
            status=status,
            error=error,
            user=os.getenv('USER'),
        )

        self.audit_log.append(entry)

        # Keep audit log size bounded (last 1000 entries)
        if len(self.audit_log) > 1000:
            self.audit_log = self.audit_log[-1000:]

    @staticmethod
    def _current_time_ms() -> int:
        """Get current time in milliseconds"""
        import time
        return int(time.time() * 1000)
